### PR TITLE
Adding the ability to export stories.

### PR DESCRIPTION
**Pull request by kiddinn**
Adding the ability in the API to export stories, and with it the first story exporter, a markdown support.

This PR adds into the API the ability to export a story into a file format.

It is modular and allows for new exporters to be written for each export format. It also has the concept of a data fetcher, possibly allowing the exporters to be used in the API client without requiring the export to happen in the backend (they'll need a different DataFetcher implemented)

### CODE REVIEW

File: timesketch/lib/stories/interface.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        elif component == 'TsViewEventList':
+            self._data_lines.append(
+                self._data_fetcher.get_view(properties.get('view')))
+        elif component == 'TsAggregationEventList':
+            self._data_lines.append(
+                self._data_fetcher.get_aggregation(
+                    properties.get('aggregation')))
+
+    def reset(self):
+        """Reset the story."""
```
Comment (berggren) [inline timesketch/lib/stories/interface.py]: Be a little more descriptive here. What is this reset used for and when?

File: timesketch/api/v1/resources.py [hunk 2]
```diff
@@ -1891,6 +1893,21 @@ def post(self, sketch_id):
 class StoryResource(ResourceMixin, Resource):
     """Resource to get a story."""
 
+    def _export_story(self, story, sketch_id, export_format='markdown'):
+        """Returns a story in a format as requested in export_format."""
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply timesketch/api/v1/resources.py]: done

File: timesketch/lib/stories/api_fetcher.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    def __init__(self):
+        """Initialize the data fetcher."""
+        super(ApiDataFetcher, self).__init__()
+        self._datastore = ElasticsearchDataStore(
+            host=current_app.config['ELASTIC_HOST'],
+            port=current_app.config['ELASTIC_PORT'])
+
+    def get_aggregation(self, agg_dict):
+        """Returns a data frame from an aggregation dict.
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply timesketch/lib/stories/api_fetcher.py]: done

File: timesketch/lib/stories/interface.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    # String representing the output format of the story.
+    EXPORT_FORMAT = ''
+
+    def __init__(self):
+        """Initialize the exporter."""
+        self._data_lines = []
+        self._data_fetcher = None
+
+    @property
+    def data(self):
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply timesketch/lib/stories/interface.py]: done

File: timesketch/lib/stories/interface.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        """Returns the data."""
+        return self.export_story()
+
+    def export_story(self):
+        """Export the story in the format."""
+        raise NotImplementedError
+
+    def from_string(self, json_string):
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply timesketch/lib/stories/interface.py]: done

File: timesketch/lib/stories/interface.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        """Feed the exporter with a JSON string."""
+        blocks = json.loads(json_string)
+        if not isinstance(blocks, (list, tuple)):
+            return
+
+        for block in blocks:
+            self.from_dict(block)
+
+    def from_dict(self, block):
+        """Feed the exporter with a single block dict."""
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply timesketch/lib/stories/interface.py]: done

File: timesketch/lib/stories/interface.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    def from_string(self, json_string):
+        """Feed the exporter with a JSON string."""
+        blocks = json.loads(json_string)
+        if not isinstance(blocks, (list, tuple)):
+            return
+
+        for block in blocks:
+            self.from_dict(block)
+
+    def from_dict(self, block):
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply timesketch/lib/stories/interface.py]: agree... done

File: timesketch/lib/stories/interface.py (hunk 1 above)
Comment (kiddinn) [reply timesketch/lib/stories/interface.py]: better?

**Comment by kiddinn** [comment]
@berggren PTAL

**Pull request merged**